Implement Writable for the Volume plugin

MikaelSmith opened the pull request:
Volumes now support writing content to files. Resolves #675.

Depends on #773 to ensure writes are immediately visible.

MikaelSmith committed:
Implement Writable for the Volume plugin

Volumes now support writing content to files.

Resolves #675.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith committed:
Refactor Docker temp container logic

Standardize and share more of the setup and teardown.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

ekinanp commented on volume/fs.go:
```diff
@@ -202,6 +202,44 @@ func (d *FS) VolumeStream(ctx context.Context, path string) (io.ReadCloser, erro
 
+// VolumeWrite satisfies the Interface required by Write to write content to a file.
+func (d *FS) VolumeWrite(ctx context.Context, path string, b []byte) error {
+	activity.Record(ctx, "Writing %v bytes to %v on %v", len(b), path, plugin.ID(d.executor))
```
Should we just log this at the `plugin.Write` level?

ekinanp commented on plugin/docker/volume.go:
```diff
@@ -110,57 +110,82 @@ func (v *volume) createContainer(ctx context.Context, cmd []string) (string, err
+type doInContainerFn = func(containerDeleter) error
+
+// Executes the callback while handling start and cleanup of an existing container.
+func (v *volume) doInTemporaryContainer(ctx context.Context, cid string, fn doInContainerFn) error {
```
Could `createContainer`'s signature be extended to `(cid, deleteFunc, error)`, and make the post-condition be that it returns a _running_ container? If so, then I don't think we need `doInTemporaryContainer`.

MikaelSmith replied:
Yeah I think so. At one point `doInTemporaryContainer` handled some of the `defer` calls, but that wasn't flexible enough.

MikaelSmith commented on volume/fs.go:
```diff
@@ -202,6 +202,44 @@ func (d *FS) VolumeStream(ctx context.Context, path string) (io.ReadCloser, erro
 
+// VolumeWrite satisfies the Interface required by Write to write content to a file.
+func (d *FS) VolumeWrite(ctx context.Context, path string, b []byte) error {
+	activity.Record(ctx, "Writing %v bytes to %v on %v", len(b), path, plugin.ID(d.executor))
```
It says pretty much the same thing in FUSE logging. This splits up path and executor id, but that's probably not important. So I think we could omit it.

ekinanp replied:
Ok, that's fine.

MikaelSmith committed:
Normalize logging

Remove logging that's redundant because the API or FUSE layer already
logs the same info. Ensure each call to Exec logs what it's executing.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith commented:
I think this is ready again.

ekinanp commented on plugin/docker/volume.go:
```diff
@@ -202,56 +206,67 @@ func (v *volume) VolumeRead(ctx context.Context, path string) ([]byte, error) {
+	}
+	defer cleanup()
+
+	// Create a tar of the file contents and upload it.
```
Why are we tar'ing and copying? Some clarification would help. Seems weird that this is different from the Kubernetes PVC implementation.

ekinanp replied:
Other than the clarification, PR looks good.

MikaelSmith replied:
Added. I suppose I could've done it with execs like everything else. That would avoid needing to pass in the permissions.

ekinanp replied:
How hard would it be to do that? The inconsistency between this and Kube PVC is kind of weird.

MikaelSmith replied:
I'll explore cleaning this up similar to Kubernetes.

I didn't add the same logic to identify an existing container because
- Docker scheduling is a lot faster, so it's not much overhead compared to starting a new process
- Docker doesn't have anything like ReadWriteOnce, so we can always get read/write access
